File: lintcode/cpp/Recover_Rotated_Sorted_Array.cpp
class Solution {
public:
    void recoverRotatedSortedArray(vector<int> &nums) {
        // write your code here
        if(nums.size() <= 1) return;
        
        int i = 1;
        for(; i < nums.size(); i++){
            if(nums[i-1] > nums[i]) break;
        }
        if(i >= nums.size()) return;
        
        reverse(nums.begin(), nums.begin()+i);
        reverse(nums.begin()+i, nums.end());
        reverse(nums.begin(), nums.end());
    }
};

File: lintcode/java/RecoverRotatedSortedArray.java
public class Solution {
    /**
     * @param nums: The rotated sorted array
     * @return: void
     */
    public void recoverRotatedSortedArray(List<Integer> nums) {
        if(nums == null || nums.size() <= 1) return;
        
        for(int i = 0; i < nums.size()-1; i++) {
            if(nums.get(i) > nums.get(i+1)) {
                reverseCollection(nums, 0, i);    
                reverseCollection(nums, i+1, nums.size()-1);
                reverseCollection(nums, 0, nums.size()-1);
                return;
            }
        }
    }
    
    private <T> void reverseCollection(List<T> nums, int i, int j) {
        while(i < j) {
            T tmp = nums.get(i);
            nums.set(i, nums.get(j));
            nums.set(j, tmp);
            i++; j--;
        }
    }
}
